kube/src/client/mod.rs

```diff
@@ -93,6 +105,37 @@ impl Client {
         Ok(res)
     }
 
+    /// Make WebSocket connection.
+    #[cfg(feature = "ws")]
+    pub async fn connect(&self, request: http::Request<()>) -> Result<WebSocketStream<ConnectStream>> {
+        let (mut parts, _) = request.into_parts();
+        if let Some(auth_header) = self.config.get_auth_header().await? {
+            parts.headers.insert(http::header::AUTHORIZATION, auth_header);
+        }
+        // Use the binary subprotocol v4, to get JSON `Status` object in `error` channel (3).
+        // There's no official documentation about this protocol, but it's described in
+        // [`k8s.io/apiserver/pkg/util/wsstream/conn.go`](https://git.io/JLQED).
+        // There's a comment about v4 and `Status` object in
+        // [`kublet/cri/streaming/remotecommand/httpstream.go`](https://git.io/JLQEh).
+        parts.headers.insert(
+            "sec-websocket-protocol",
+            "v4.channel.k8s.io".parse().expect("valid header value"),
+        );
+        // Replace scheme to ws(s).
+        let pandq = parts.uri.path_and_query().expect("valid path+query from kube");
+        parts.uri = finalize_url(&self.cluster_url, &pandq)
+            .replacen("http", "ws", 1)
+            .parse()
+            .expect("valid URL");
+        let (stream, _) = connect_async_with_tls_connector(
+            http::Request::from_parts(parts, ()),
+            Some(self.tls_connector.clone()),
+        )
+        .await
+        .expect("handshake error");
```

Comment:
I hit this panic when my service account didn't have sufficient permission to run the `exec` and the API server returned a 403. IMO returning an error would be better in this case.

Reply:
yeah, you can also hit this if the pod you are trying to attach to does not exist

Reply:
@jbg Thanks for trying and catching this. I had noticed it while testing, but forgot about it. I changed to return `Error::Api` instead.